shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -55,63 +85,103 @@ public void validateArgs(CommandLine cl) throws InvalidArgumentException {
-    if (status.isFolder()) {
-      List<URIStatus> statuses = mFileSystem.listStatus(filePath);
-      List<String> errorMessages = new ArrayList<>();
-      for (URIStatus uriStatus : statuses) {
-        AlluxioURI newPath = new AlluxioURI(uriStatus.getPath());
-        try {
-          persist(newPath);
-        } catch (Exception e) {
-          errorMessages.add(e.getMessage());
-        }
-      }
-      if (errorMessages.size() != 0) {
-        throw new IOException(Joiner.on('\n').join(errorMessages));
-      }
-    } else if (status.isPersisted()) {
-      System.out.println(filePath + " is already persisted");
-    } else {
-      try {
-        FileSystemUtils.persistFile(mFileSystem, filePath);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        throw new RuntimeException(e);
-      } catch (TimeoutException e) {
-        throw new RuntimeException(e);
+    mTotalFilesToPersist = mToPersist.size();
+    System.out.println("Found " + mTotalFilesToPersist + " files to persist.");
+    ExecutorService service = Executors.newFixedThreadPool(parallelism);
+    List<Future<Void>> persistCalls = new ArrayList<>(parallelism);
+    for (int i = 0; i < parallelism; i++) {
+      persistCalls.add(service.submit(new PersistCallable()));
+    }
+    try {
+      for (Future<Void> call : persistCalls) {
+        call.get();
       }
-      System.out.println("persisted file " + filePath + " with size " + status.getLength());
+    } catch (ExecutionException e) {
+      System.out.println("Fatal error: " + e);
+    } catch (InterruptedException e) {
+      System.out.println("Persist interrupted, exiting.");
     }
+    return 0;
   }
 
   @Override
   public String getUsage() {
-    return "persist <path> [<path> ...]";
+    return "persist [--parallelism <#>] <path> [<path> ...]";
   }
 
   @Override
   public String getDescription() {
     return "Persists files or directories currently stored only in Alluxio to the "
         + "UnderFileSystem.";
   }
+
+  /**
+   * Clears the queue to persist and internal variables. This class is not designed to be used
+   * concurrently or from an existing incomplete state.
+   */
+  private void resetState() {
+    mToPersist.clear();
+    mFilesPersisted.set(0);
+    mTotalFilesToPersist = 0;
+  }
+
+  /**
+   * Thread that polls the persist queue and persists files.
+   */
+  private class PersistCallable implements Callable<Void> {
+    @Override
+    public Void call() throws Exception {
+      AlluxioURI toPersist = mToPersist.poll();
+      while (toPersist != null) {
+        try {
+          FileSystemUtils.persistFile(mFileSystem, toPersist);
+          // There is a slight chance of out of order output to the client.
+          // If this becomes a problem we can consider locking instead of an atomic int.
+          String progress =
+              "(" + mFilesPersisted.incrementAndGet() + "/" + mTotalFilesToPersist + ")";
```

Comment:
I first test persist 9 files with parallelism as 6 and then test persist 57 files with parallelism as 6. All the test has out of order issue, especially in 57 files test.

I feel that it will be better if
1. don't show the progress numbers. In the end, show how many files have been successfully persisted. If the number of files is huge, we could show progress number as the multiple of parallelism factor.  (6/57) , (12/57), ... (54, 57), (57, 57) 
2. add locking to make sure the number is sequential.

Reply:
3. Have a single thread write to stdout, and pass messages to it through a queue as persist jobs succeed/fail

Reply:
Thanks for testing, I wasn't able to get out of order execution when I tried. I added a lock for the progress.